src/storage/src/controller.rs

```diff
@@ -1006,14 +1221,90 @@ where
             persist: persist_clients,
         }
     }
+}
 
+impl<T> Controller<T>
+where
+    T: Timestamp + Lattice + TotalOrder + TryInto<i64> + TryFrom<i64> + Codec64 + Unpin,
+    <T as TryInto<i64>>::Error: std::fmt::Debug,
+    <T as TryFrom<i64>>::Error: std::fmt::Debug,
+
+    // Required to setup grpc clients for new storaged instances.
+    StorageCommand<T>: RustType<ProtoStorageCommand>,
+    StorageResponse<T>: RustType<ProtoStorageResponse>,
+
+    Self: StorageController<Timestamp = T>,
+{
     /// Validate that a collection exists for all identifiers, and error if any do not.
-    fn validate_ids(&self, ids: impl Iterator<Item = GlobalId>) -> Result<(), StorageError> {
+    fn validate_collection_ids(
+        &self,
+        ids: impl Iterator<Item = GlobalId>,
+    ) -> Result<(), StorageError> {
         for id in ids {
             self.collection(id)?;
         }
         Ok(())
     }
+
+    /// Validate that a collection exists for all identifiers, and error if any do not.
+    fn validate_export_ids(&self, ids: impl Iterator<Item = GlobalId>) -> Result<(), StorageError> {
+        for id in ids {
+            self.export(id)?;
+        }
+        Ok(())
+    }
+
+    // Should only fail if collection doesn't exist. N.B. We can't just take in the mut ref because then the borrow checker wouldn't let us read state.
+    fn generate_new_capability_for_collection<F>(
+        &mut self,
+        id: GlobalId,
+        f: F,
+    ) -> Result<ChangeBatch<<Self as StorageController>::Timestamp>, StorageError>
+    where
+        F: FnOnce(&mut CollectionState<<Self as StorageController>::Timestamp>),
+    {
+        let collection = self
+            .state
+            .collections
+            .get_mut(&id)
+            .ok_or(StorageError::IdentifierMissing(id))?;
+        f(collection);
+
+        let mut update = ChangeBatch::new();
+
+        // Get read policy sent from the coordinator
+        let mut new_read_capability = collection
+            .read_policy
+            .frontier(collection.write_frontier.frontier());
+
+        // Also consider the write frontier of any exports
+        for export_id in self
+            .state
+            .exported_collections
+            .get(&id)
+            .cloned()
+            .unwrap_or_default()
+        {
+            new_read_capability.meet_assign(
```

Comment:
Don't we have to make sure that we never allow the `since` of depended-upon collections to advance "up to" the write frontier of the sink? So that on restart we can still differentiate between records that need to be written and records that don't need to be written? I know you saw in your tests that you had duplicate records because of this bug so you must have fixed it somehow. I don't immediately see where this is fixed, though. :sweat_smile:

Reply:
yes, that bug was fixed by computing the as_of using the since of the collection above!

I'll double check but I believe that this is an invariant we already have by nature of how the write_frontier is reported from the sink.